src/executor/limit.rs

```diff
@@ -1,52 +1,49 @@
 // Copyright 2022 RisingLight Project Authors. Licensed under Apache-2.0.
 
 use super::*;
-use crate::array::DataChunk;
+use crate::array::{DataChunk, DataChunkBuilder};
+use crate::types::DataType;
 
 /// The executor of a limit operation.
 pub struct LimitExecutor {
     pub child: BoxedExecutor,
     pub offset: usize,
     pub limit: usize,
+    pub output_types: Vec<DataType>,
 }
 
 impl LimitExecutor {
     #[try_stream(boxed, ok = DataChunk, error = ExecutorError)]
     pub async fn execute(self) {
         // the number of rows have been processed
         let mut processed = 0;
-        let mut dummy_chunk = None;
+        let mut builder = DataChunkBuilder::new(self.output_types.iter(), PROCESSING_WINDOW_SIZE);
 
         #[for_await]
         for batch in self.child {
-            let batch = batch?;
-            if dummy_chunk.is_none() {
-                dummy_chunk = Some(batch.slice(0..0));
-                if self.offset == 0 && self.limit == 0 {
-                    break;
-                }
+            if self.limit == 0 {
+                break;
             }
+            let batch = batch?;
             let cardinality = batch.cardinality();
             let start = processed.max(self.offset) - processed;
             let end = (processed + cardinality).min(self.offset + self.limit) - processed;
             processed += cardinality;
             if start >= end {
                 continue;
             }
-            if (start..end) == (0..cardinality) {
-                yield batch;
-            } else {
-                yield batch.slice(start..end);
+            for row in batch.rows().skip(start).take(end - start) {
+                if let Some(chunk) = builder.push_row(row.values()) {
+                    yield chunk
+                }
```

Comment:
Iterating over batch here seems to introduce addtional overheads. Maybe we just don't need to split chunks in limit?

Reply:
I think some executors need to split chunks because they iterate over the inputs row by row and then generate different distribution of data, e.g., sort, join. The inputs are splitted, but the output is a single big chunk.

For limit, it only forwards input chunks, so no addtional effort to split chunks is needed.